### DISCUSSION TITLE
When I zoom by chartjs-plugin-zoom library, the children of tree move on the screen, and the zooming speed is very slow in the mobile size.

### DISCUSSION THREAD

**Discussion by mirzaeianfar** [kurkle/chartjs-chart-treemap#155] [Q&A]
When I zoom by chartjs-plugin-zoom library, the children of tree move on the screen, and the zooming speed is very slow in the mobile size.
How can I avoid moving  the items of tree  and touch speed more in mobile size?

![Untitled](https://user-images.githubusercontent.com/53333713/196425032-4a5098f6-77a3-4712-9a2e-6b644e5d0da3.png)

**Comment by mirzaeianfar** [comment]
You can test it in the address:
https://bv.emofid.com/map
@stockiNail

**Reply by stockiNail** [reply]
I had a look but the treemap controller is correctly updating itself at every zoom (the plugin is working in this way). Maybe you can play with `speed` zoom plugin option if you haven't done yet. 
At controller level, we need to have a look to the following call if we can optimized in same way, but it shouldn't be invoked when you are zooming by wheel.

https://github.com/kurkle/chartjs-chart-treemap/blob/4a0b22a718b4da4146c28e1df4e5b13f16edd337/src/controller.js#L125

**Reply by stockiNail** [reply]
The call that I mentioned is already cached, therefore nothing to that direction.

**Reply by mirzaeianfar** [reply]
The speed zoom config is for wheel on screen but for pinch config in mobile device does not work.
We do not have speed config for pinch config in mobile device.
Also when we zoom by pinch in mobile device, items move an other place so we do not like to move items or rectangle small items.
Zooming in mobile devices is very slow.
Please test it in mobile devices with old mobile phone.
Can you help me more?
Thank you so much

**Reply by kurkle** [reply]
I think this could be optimized by not doing layout if the aspect ratio stays the same. Probably requires some refactoring.

**Reply by stockiNail** [reply]
@kurkle when you write " if the aspect ratio stays the same", what exactly do you mean?

**Reply by stockiNail** [reply]
I'm doing some tests, checking if the `tree` is changed.

**Reply by kurkle** [reply]
> @kurkle when you write " if the aspect ratio stays the same", what exactly do you mean? 

The layout needs to be redone on resize and ofc when data changes. I thought of the resize case, but this is actually only scale change where rects should keep shape and order, maybe even when only one of the scales is zoomed?

**Reply by stockiNail** [reply]
Correct, here only the min/max of the scale are changed. If we don't redo the layout, the elements will be drawn with the previous coordinates and therefore nothing changed because the chart area remains the same.

**Reply by kurkle** [reply]
Thats why some refactoring is needed :)

Something like

Layout: calculate rect positions and sizes on a normalized (1x1) container rectangle

Transform: (also when scales change) update those normalized values with getPixelForValue

Edit: the normalized rect needs to be based on chartArea (aspect ratio matters)

Edit2: actually datasetArea or ehat was it called, in case the chartArea is divided by stacked scales

**Reply by stockiNail** [reply]
If understood well, we could always calculate the sizes based on chartArea (as it was before) and transform the canvas before drawing based on new coordinates of new scale size

**Reply by stockiNail** [reply]
It works. Some options are not working (like overflow) because we just resizing the canvas, always with the same size. I need to do some additional tests. Let me move the discussion in a issue

**Reply by stockiNail** [reply]
@kurkle there is a bug in the `arrayNotEquals` function: 

https://github.com/kurkle/chartjs-chart-treemap/blob/4a0b22a718b4da4146c28e1df4e5b13f16edd337/src/controller.js#L20

wrong array option: `a1.lenght`

**Reply by stockiNail** [reply]
@mirzaeianfar I have submitted a PR  #157. I'm not able to understand if this new implementation is really improving your use case because I don't have old mobile to test. If I provide the new controller JS file (in some way), are you able to test in the mobile?

**Reply by stockiNail** [reply]
@mirzaeianfar I’m not 100% sure that the slowness depends on update phase. Could you provide the whole configuration of the chart, mainly the dataset one ? I think you are using scriptable options which could affect the draw phase.

**Comment by kurkle** [comment]
@mirzaeianfar can you test with version 2.1.3 and see if it is better and if it is good enough for your use case?
